Flaky TestFrontendStreamingResponse

`TestFrontendStreamingResponse` is flaky ([CI](https://github.com/grafana/mimir/actions/runs/9895618542/job/27362476978?pr=8688)):

```
--- FAIL: TestFrontendStreamingResponse (0.07s)
    --- FAIL: TestFrontendStreamingResponse/context_cancelled_while_streaming_response (0.01s)
        frontend_test.go:508: 
            	Error Trace:	/__w/mimir/mimir/pkg/frontend/v2/frontend_test.go:508
            	Error:      	Not equal: 
            	            	expected: "part 1/3"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-part 1/3
            	            	+
            	Test:       	TestFrontendStreamingResponse/context_cancelled_while_streaming_response
```

another instance

```
--- FAIL: TestFrontendStreamingResponse (0.07s)
    --- FAIL: TestFrontendStreamingResponse/context_cancelled_while_streaming_response (0.02s)
        frontend_test.go:508: 
            	Error Trace:	/__w/mimir/mimir/pkg/frontend/v2/frontend_test.go:508
            	Error:      	Not equal: 
            	            	expected: "part 1/3"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-part 1/3
            	            	+
            	Test:       	TestFrontendStreamingResponse/context_cancelled_while_streaming_response
level=info msg="adding connection to query-scheduler" addr=127.0.0.1:34041
level=debug user=test msg="enqueuing request"
method=frontendSchedulerWorker.enqueueRequest user=test level=warn msg="scheduler reported it has too many outstanding requests"
level=debug user=test msg="request enqueued successfully, waiting for response"
level=debug user=test msg="received response"
level=debug user=test msg="enqueuing request"
method=frontendSchedulerWorker.enqueueRequest user=test level=warn msg="received error while receiving response" err="rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"max_age\""
level=debug user=test msg="enqueuing request failed, will retry"
level=debug user=test msg="enqueuing request"
level=error msg="error sending requests to scheduler" err="rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"max_age\"" addr=127.0.0.1:34041
method=frontendSchedulerWorker.enqueueRequest user=test level=warn msg="scheduler reported it has too many outstanding requests"
level=debug user=test msg="request enqueued successfully, waiting for response"
level=debug user=test msg="received response"
level=debug msg="stream context finished" err="context canceled"
FAIL
```

--- a/pkg/frontend/v2/frontend_test.go
+++ b/pkg/frontend/v2/frontend_test.go
@@ -450,16 +450,17 @@ func TestFrontendStreamingResponse(t *testing.T) {
 			name: "context cancelled while streaming response",
 			sendResultStream: func(f *Frontend, msg *schedulerpb.FrontendToScheduler) error {
 				ctx, cancelCause := context.WithCancelCause(user.InjectOrgID(context.Background(), userID))
-				recvCalled := make(chan struct{})
+				recvCalled := make(chan chan struct{})
 				cancelAfterCalls := 2
 				s := &mockQueryResultStreamServer{ctx: ctx, queryID: msg.QueryID, recvCalled: recvCalled}
 				go func() {
 					recvCount := 0
-					for range recvCalled {
+					for called := range recvCalled {
 						recvCount++
 						if recvCount == cancelAfterCalls {
 							cancelCause(fmt.Errorf("streaming cancelled"))
 						}
+						called <- struct{}{}
 					}
 				}()
 				s.msgs = append(s.msgs,
@@ -539,7 +540,7 @@ type mockQueryResultStreamServer struct {
 	queryID    uint64
 	msgs       []*frontendv2pb.QueryResultStreamRequest
 	next       int
-	recvCalled chan struct{}
+	recvCalled chan chan struct{}
 
 	grpc.ServerStream
 }
@@ -553,14 +554,16 @@ func (s *mockQueryResultStreamServer) SendAndClose(_ *frontendv2pb.QueryResultRe
 }
 
 func (s *mockQueryResultStreamServer) Recv() (*frontendv2pb.QueryResultStreamRequest, error) {
-	if s.recvCalled != nil {
-		s.recvCalled <- struct{}{}
-	}
 	if err := s.ctx.Err(); err != nil {
 		if s.recvCalled != nil {
 			close(s.recvCalled)
 		}
 		return nil, err
+	}
+	if s.recvCalled != nil {
+		called := make(chan struct{}, 1)
+		s.recvCalled <- called
+		<-called
 	}
 	if s.next >= len(s.msgs) {
 		return nil, io.EOF

